Add resource properties structure and validation (#1213)
awslabs/smithy: docs/source/1.0/spec/core/model.rst

- Child resources MUST provide identifier bindings for all of its parent's
  identifiers.
- Identifier bindings are formed on input or output structure members that are

>   marked as :ref:`required <required-trait>`.

Author: I agree this needs correction. I think a more exact description (not RST formatted) would be:
```
Identifier bindings may be formed on input or output structure members that are marked as required.
```

While output properties don't _need_ it, identifier bindings on output are a filter for properties on output members. A create lifecycle operation (or likely any collection operation) is likely to have an output that is a resource identifier and implicitly so. If they're marked as required (the service always returns them) there's no further trait markup needed to remove them from resource properties consideration.

to the "historicalId" identifier.

.. _resource-properties:

Resource Properties
====================

:dfn:`Resource properties` represent the state of a resource within a service.
Properties can be referred to in the top level input and output shapes
of a resource's instance operations, including create, read, update,
delete, and put. All declared resource properties MUST appear in at
least one instance operation's input or output.

For example, the following model defines a ``Forecast`` resource with a
single property ``chanceOfRain`` read by the GetForecast operation, and

.. tabs::

    .. code-tab:: smithy

        $version: "2.0"
        namespace smithy.example

        resource Forecast {
            properties: { chanceOfRain: Float }
            read: GetForecast
        }

        @readonly
        operation GetForecast {
           output: GetForecastOutput
        }

>         structure GetForecastOutput {

Reviewer: use type elision syntax

            read: GetForecast
        }

        @readonly
        operation GetForecast {
           output: GetForecastOutput
        }

        structure GetForecastOutput {
            chanceOfRain: Float
        }

    .. code-tab:: json

        {
            "smithy": "2.0",
            "shapes": {
                "smithy.example#Forecast": {
                    "type": "resource",
                    "read": {
                        "target": "smithy.example#GetForecast"
                    }
                },
                "smithy.example#GetForecast": {
                    "type": "operation",
                    "output": {
                        "target": "smithy.example#GetForecastOutput"
                    },
                    "traits": {
                        "smithy.api#readonly": {}
                    }
                },
                "smithy.example#GetForecastOutput": {
                    "type": "structure",
                    "members": {
                        "chanceOfRain": {
                            "target": "smithy.api#Float"
                        }
                    },
                }
            }
        }

> The:ref:`property-trait` can be used, while specifying the `name` property, to

Reviewer: ```suggestion
The `name` property of the :ref:`property-trait` is used to
```

                        "target": "smithy.example#GetForecastOutput"
                    },
                    "traits": {
                        "smithy.api#readonly": {}
                    }
                },
                "smithy.example#GetForecastOutput": {
                    "type": "structure",
                    "members": {
                        "howLikelyToRain": {
                            "target": "smithy.api#Float",
                            "traits": {
                                "smithy.api#property": {
                                    "name": "chanceOfRain"
                                }
                            }
                        }
                    }
                }
            }
        }

.. rubric:: Resource property binding validation

- All top-level input or output members must bind to a resource property unless
  marked with :ref:`notProperty-trait` or another trait with it applied.
- Top-level members of the input and output of resource instance operations MUST
  only use properties that resolve to declared resource properties except for
  members marked with the @notProperty trait or marked with traits marked with

>   the @notProperty trait.

Author: Fixed in multiple places here.
